usl_pipeline/study_area_uploader/tests/chunkers/test_study_area_chunkers.py

```diff
@@ -0,0 +1,255 @@
+import pathlib
+import shutil
+import tarfile
+import tempfile
+from typing import Optional, Tuple
+import typing.io
+from unittest import mock
+
+from google.cloud import storage
+import numpy
+from numpy import testing
+import rasterio
+from shapely import geometry
+
+from study_area_uploader.chunkers import study_area_chunkers
+from study_area_uploader.transformers import study_area_transformers
+from usl_lib.readers import elevation_readers, polygon_readers
+from usl_lib.shared import geo_data
+from usl_lib.storage import file_names
+
+
+def prepare_test_geotiff_elevation_file(
+    data: list[list[float]],
+    file_path: pathlib.Path,
+    x_ll_corner: float = 0.0,
+    y_ll_corner: float = 0.0,
+    cell_size: float = 1.0,
+) -> None:
+    """Creates a test GeoTIFF elevation file."""
+    height = len(data)
+    width = len(data[0])
+    with rasterio.open(
+        str(file_path),
+        "w",
+        driver="GTiff",
+        dtype=rasterio.float32,
+        nodata=0.0,
+        width=width,
+        height=height,
+        count=1,
+        crs=rasterio.CRS.from_epsg(32618),
+        transform=rasterio.Affine(
+            cell_size,
+            0.0,
+            x_ll_corner,
+            0.0,
+            -cell_size,
+            y_ll_corner + cell_size * height,
+        ),
+    ) as raster:
+        raster.write(numpy.array(data).astype(rasterio.float32), 1)
+
+
+def bbox_polygon(x1: float, y1: float, x2: float, y2: float) -> geometry.Polygon:
+    """Converts a bounding box defined by two opposite corners to a Polygon."""
+    return geometry.Polygon([(x1, y1), (x2, y1), (x2, y2), (x1, y2), (x1, y1)])
+
+
+def copy_stream_to_file(
+    input_fd: typing.BinaryIO,
+    output_file_path: pathlib.Path,
+) -> None:
+    """Copies file stream to local file (wrapper around shutil.copyfileobj)."""
+    with open(output_file_path, "wb") as output_fd:
+        shutil.copyfileobj(input_fd, output_fd)
+
+
+def mock_file_blob(work_dir: pathlib.Path, relative_path: str) -> storage.Blob:
```

Comment:
for testing the
```python
        study_area_chunk_bucket.blob(
            f"{study_area_name}/{chunk_file_name}.tar"
        ).upload_from_file(tar_fd)
```
call, what I'd do is
```python
mock_bucket.blob.assert_has_calls([
  mock.call('study_area_name/chunk_1.tar'),
  mock.call('study_area_name/chunk_2.tar'),
])
tar_uploaded_for_chunk_1 = mock_bucket.blob().upload_from_file.mock_calls[0].args[0]
tar_uploaded_for_chunk_2 = mock_bucket.blob().upload_from_file.mock_calls[1].args[0]
```
and then you can assert properties of the tar you uploaded.

Reply:
Done